Feature/client side encryption: internal design doc review (#1)
cloudogu/velero: design/client-side-encryption_design.md

# Proposal for client-side encryption

## Abstract

Currently only PersistentVolumes are stored in a Restic/Kopia repository, which means that only they are encrypted and the rest of the backup is not.
It is desired that the cluster backup can be encrypted which overall will lead to better security.

## Background

As of today, Velero stores only PersistentVolume data in a Restic/Kopia repository.
This repository is encrypted while the rest of the backup is not.
Backups should be safe against illegal access and tampering.
It is therefore desired to have the ability to store this data in an encrypted manner.

## Goals

- Encryption of the main cluster backup file with AES.
- Basic configuration of the encryption, e.g. turning it on and off.

> - Changing the encryption key for future backups while still retaining access to older backups.

Author: maybe too long and `changing the encryption key` as a goal is a bit misleading.
```suggestion
- Ability to change encryption key for future backups
```

> ## Non Goals
> 
> - Other ciphers than AES (can be added later).
> - Encrypting anything other than the main backup file.

Author: ```suggestion
## Non Goals

- Other ciphers than AES (can be added later).
- Encrypting anything other than the main backup file.
- re-encrypt old backups if new encryption key was set
```

Author: Maybe also add a non-goal of supporting other ways to get the encryption key for now, e.g. Hashicorp Vault or others

## High-Level Design

Encryption will happen during the backup process before the objects are persisted (e.g. written to an object store).
Decryption will happen during a restore after the persisted objects have been read.
The encryption key will be stored in a Kubernetes secret.
Encryption can be configured through commandline arguments to the velero server.

> Changing the encryption key is done by creating a new secret with the changed encryption key and configuring it in Velero's configuration.

Author: As i ask myself in the section below, who is responsible for creating the secret. Maybe make this more clear in this part.
```suggestion
The encryption key will be stored in a Kubernetes secret by the user.
Encryption can be configured through commandline arguments to the velero server.
The user can change the encryption key by creating a new encryption key secret and configuring it in Velero's configuration.
```

If a backup is encrypted, it contains a flag and the name of the used encryption key secret in its metadata.

> For backwards compatability reasons, only the main backup archive will be encrypted.

Author: ```suggestion
For backwards compatibility reasons, only the main backup archive will be encrypted.
```

## Detailed Design

### Configuration

To enable encryption, a feature flag is introduced:
```go
// EncryptionFeatureFlag is the feature flag string that defines whether or not client-side-encryption features are being used.
const EncryptionFeatureFlag = "EnableEncryption"
```

A commandline flag is added to the `velero server` command to set the name of the secret containing the encryption key:
```go
type serverConfig struct {
    ...
	encryptionSecret string
}
```

```go
command.Flags().StringVar(&config.encryptionSecret, "encryption-secret", config.encryptionSecret, "Name of the secret that contains the key to encrypt backups with. Used only if the 'EnableEncryption' feature flag is set.")

> ```

Author: Does the user always have to create the secret by themself?

Reviewer: Currently yes, but I mentioned it under open issues, that the encryption key could be generated.

### En-/Decryption implementation

Encryption is located in the `encryption` package.
The `encryptor` interface has been introduced to provide abstraction over the used encryption standard.
The default (and for now only) encryptor implementation is the `aesEncryptor`:

> ```go

Author: Is it ok to add method bodies and imports to the design document? I think, this is not necessary to agree on names of types etc.

Reviewer: Yeah, maybe not necessary, but I want readers to understand what's happening.

	}

	buf := new(bytes.Buffer)
	_, err = io.Copy(buf, in)
	if err != nil {
		return nil, fmt.Errorf("failed to copy input to buffer: %w", err)
	}

	ciphertext := buf.Bytes()
	plaintext, err := encryptor.decrypt(ciphertext)
	if err != nil {
		return nil, err
	}

	return bytes.NewReader(plaintext), nil
}
```

### Storing the encryption status

To discern if and how a given backup is encrypted, the `BackupStatus` gets extended by an `Encryption` field of type `EncryptionStatus`:
```go
// BackupStatus captures the current status of a Velero backup.
type BackupStatus struct {
    ...
    
    // Encryption contains metadata about and whether encryption was used.
    // +optional
    Encryption EncryptionStatus `json:"encryption,omitempty"`
}

// EncryptionStatus contains information about the encryption of a backup.
type EncryptionStatus struct {
	// IsEncrypted indicates whether this backup is encrypted.
	IsEncrypted bool `json:"isEncrypted,omitempty"`
	// EncryptionSecret is the name of the secret containing the encryption key used for encryption.
	EncryptionSecret string `json:"encryptionSecret,omitempty"`
}
```

### Encryption on backup

> The `kubernetesBackupper` gets extended by a namespace and the encryption secret:

Author: If any other future feature needs a namespace, i think we should not reuse the field as this would couple the secret location to whatever else is needed. To make that clear you could consider renaming it to `encryptionSecretNamespace`.

Author: For the possible future support of other methods to get the encryption key, you could consider an interface `encryptionKeyReceiver`

Reviewer: I renamed  `namespace` to `veleroNamespace`.

Reviewer: Implemented a `KeyRetriever` interface like you asked.

```go
type kubernetesBackupper struct {
	...
	namespace        string
	encryptionSecret string
}

func NewKubernetesBackupper(
	...
    namespace string,
    encryptionSecret string,
) (Backupper, error) {
	...
}
```

The namespace is the namespace velero runs in.
It is necessary to find the encryption secret.

Encryption happens in the `BackupWithResolvers` method:
```go
func (kb *kubernetesBackupper) BackupWithResolvers(log logrus.FieldLogger,
	backupRequest *Request,
	backupFile io.Writer,
	backupItemActionResolver framework.BackupItemActionResolverV2,
	volumeSnapshotterGetter VolumeSnapshotterGetter) (backupErr error) {

	var backupContent io.Writer
	if features.IsEnabled(velerov1api.EncryptionFeatureFlag) {

>         encryptionKey, err := encryption.GetEncryptionKeyFromSecret(kb.kbClient, kb.encryptionSecret, kb.namespace)

Author: Please fix Indentation

Since only the main backup archive is encrypted, it should be documented precisely what information gets encrypted and what not.
Unencrypted information includes:
- CSI-VolumeSnapshotClasses
- CSI-VolumeSnapshotContents (despite the name, these do not contain the contents of the backup)
- CSI-VolumeSnapshots
- Item-Operations
- Backup-Logs
- PodVolumeBackups
- a resource list containing the names and kinds of the backed up resources
- Results (errors, warnings) of the backup operation
- (native) VolumeSnapshots
- the velero backup custom resource

If the secrets containing the encryption keys are deleted, users might lose access to their backups.
Therefore, users should back up their encryption keys outside the cluster.

## Compatibility

Encrypting the whole backup would introduce major breaking changes for object store plugins and maybe even more.
Therefore, only the main backup archive is encrypted.

If unencrypted backups exist when the feature is activated, they will stay unencrypted and can still be restored.
Likewise, when the feature gets disabled, encrypted backups can still be decrypted.

> It is also possible to change the encryption key without loosing access to previous backups.

Author: ```suggestion
It is also possible to change the encryption key without losing access to previous backups.
```

This is due to the backups containing metadata about their encryption.
If this metadata is absent, the backup is assumed to be unencrypted.
Currently, this encryption status contains a flag whether the backup is encrypted, as well as the name of the secret containing the key it was encrypted with.

> In the future, the encryption status can be extended by the type of encryption without loosing backwards compatability by assuming the default to be AES.

Author: ```suggestion
In the future, the encryption status can be extended by the type of encryption without losing backwards compatibility by assuming the default to be AES.
```

There should be no breaking changes.

## Implementation

A working prototype already exists.
Therefore, minimal effort is needed to implement this.

> Timelines are almost entirely dependent on how fast this goes through and how many things need changing before it can.

Author: This sentence seems to be unfinished.
```suggestion
Timelines are almost entirely dependent on how fast this goes through and how many things need to be changed before it can be merged.
```

    return SecretKeyConfig(s.secretName, s.namespace)
}

// GetKey fetches an encryption key from a Kubernetes secret.
func (s *secretKeyRetriever) GetKey() (string, error) {
    secret := corev1.Secret{}
    err := s.client.Get(context.Background(), crlClient.ObjectKey{Name: s.secretName, Namespace: s.namespace}, &secret)
    if err != nil {
        return "", fmt.Errorf("failed to get encryption key secret '%s': %w", s.secretName, err)
    }
    
    key, ok := secret.Data[encryptionKeySecretField]
    if !ok {
        return "", fmt.Errorf("encryption key secret '%s' lacks field '%s'", s.secretName, encryptionKeySecretField)
    }
    
    return string(key), nil
}

// SecretKeyConfig creates the retriever config for a key retriever that fetches the encryption key from a secret.
func SecretKeyConfig(secretName, namespace string) velerov1.EncryptionKeyRetrieverConfig {
    return velerov1.EncryptionKeyRetrieverConfig{
        locationSecretNameKey: secretName,
        locationNamespaceKey:  namespace,
    }
}
```

### Encryption on backup

The `kubernetesBackupper` gets extended by a namespace and the encryption secret:
```go
type kubernetesBackupper struct {
	...

> 	veleroNamespace        string

Author: fix identation

	encryptionSecret string
}

func NewKubernetesBackupper(
	...
    veleroNamespace string,
    encryptionSecret string,
) (Backupper, error) {
	...
}
```

The namespace is the namespace velero runs in.
It is necessary to find the encryption secret.

Encryption happens in the `BackupWithResolvers` method:
```go
func (kb *kubernetesBackupper) BackupWithResolvers(log logrus.FieldLogger,
	backupRequest *Request,
	backupFile io.Writer,
	backupItemActionResolver framework.BackupItemActionResolverV2,
	volumeSnapshotterGetter VolumeSnapshotterGetter) (backupErr error) {

    var backupContent io.Writer
    if features.IsEnabled(velerov1api.EncryptionFeatureFlag) {
        encryptionKeyRetriever, err := encryption.KeyRetrieverFor(
			encryption.SecretKeyRetrieverType,
            encryption.SecretKeyConfig(kb.encryptionSecret, kb.veleroNamespace),
            kb.kbClient,

>         )

Author: fix identation of the parameters
